src/dx/presence/mod.rs

```diff
@@ -0,0 +1,547 @@
+    /// # use std::sync::Arc;
+    /// let mut pubnub = // PubNubClient
+    /// #         PubNubClientBuilder::with_reqwest_transport()
+    /// #             .with_keyset(Keyset {
+    /// #                 subscribe_key: "demo",
+    /// #                 publish_key: None,
+    /// #                 secret_key: None
+    /// #             })
+    /// #             .with_user_id("uuid")
+    /// #             .build()?;
+    /// pubnub
+    ///     .set_state(HashMap::<String, bool>::from(
+    ///          [("is_admin".into(), false)]
+    ///      ))
+    ///     .channels(["lobby".into(), "announce".into()])
+    ///     .channel_groups(["area-51".into()])
+    ///     .execute()
+    ///     .await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn set_state<S>(&self, state: S) -> SetStateRequestBuilder<T, D>
+    where
+        S: Serialize + Send + Sync + 'static,
+    {
+        SetStateRequestBuilder {
+            pubnub_client: Some(self.clone()),
+            state: Some(state.serialize().ok()),
+            user_id: Some(self.config.user_id.clone().to_string()),
+            ..Default::default()
+        }
+    }
+
+    /// Create a heartbeat request builder.
+    ///
+    /// This method is used to update state associated with `user_id` on
+    /// channels using `heartbeat` operation endpoint.
+    ///
+    /// Instance of [`HeartbeatRequestsBuilder`] returned.
+    pub fn set_state_with_heartbeat<U>(&self, state: U) -> HeartbeatRequestBuilder<T, D>
+    where
+        U: Serialize + Send + Sync + 'static,
+    {
+        self.heartbeat().state(state)
+    }
+}
+
+impl<T, D> PubNubClientInstance<T, D>
+where
+    T: Transport + Send + 'static,
+    D: Deserializer + 'static,
+{
+    /// Announce `join` for `user_id` on provided channels and groups.
+    #[cfg(feature = "std")]
+    #[allow(dead_code)]
+    pub(crate) fn announce_join(
+        &self,
+        channels: Option<Vec<String>>,
+        channel_groups: Option<Vec<String>>,
+    ) {
+        self.configure_presence();
+
+        {
+            let slot = self.presence.read();
+            if let Some(presence) = slot.as_ref() {
+                presence.announce_join(channels, channel_groups);
+            }
+        };
+    }
+
+    /// Announce `leave` for `user_id` on provided channels and groups.
+    #[cfg(feature = "std")]
+    #[allow(dead_code)]
+    pub(crate) fn announce_left(
+        &self,
+        channels: Option<Vec<String>>,
+        channel_groups: Option<Vec<String>>,
+    ) {
+        self.configure_presence();
+
+        {
+            let slot = self.presence.read();
+            if let Some(presence) = slot.as_ref() {
+                presence.announce_left(channels, channel_groups);
+            }
+        };
+    }
+
+    /// Complete presence configuration.
+    ///
+    /// Presence configuration used only with presence event engine.
+    #[cfg(feature = "std")]
+    #[allow(dead_code)]
```

Comment:
Why there is a dead_code? It seems like a needed utility.
Or is it needed by subscribe in the next PR?

Reply:
It was part of a commit that was not pushed before the code was finished.